qutip/solver/heom/bofin_baths.py

```diff
@@ -337,6 +346,188 @@ def combine(cls, exponents, rtol=1e-5, atol=1e-7):
 
         return new_exponents
 
+    def spectral_density(self, w):
+        """
+        Spectral density of this bath.
+
+        The BosonicBath class mainly represents a list of expansion
+        coefficients of the bath auto-correlation function. However, subclasses
+        may additionally override this `spectral_density` function to calculate
+        the bath spectral density from a known analytic expression.
+        Implementing this function does not affect the HEOM simulation, but it
+        can be useful for illustration purposes. For example, the
+        `correlation_function` numerically computes the correlation function
+        based on the provided `spectral_density` function. It can be compared
+        to `correlation_function_approx`, which returns the correlation
+        function according to the list of expansion coefficients that is used
+        for the HEOM construction.
+
+        Note that various conventions for the definition of the spectral
+        density exist. We follow the convention described in Eq. 7 of the BoFiN
+        paper (DOI: 10.1103/PhysRevResearch.5.013181). We also assume that
+        the spectral density is zero at negative frequencies (otherwise, no
+        thermal state exists).
+        """
+
+        raise ValueError(
+            "The spectral density of this bath has not been specified")
+
+    def correlation_function(self, t, **kwargs):
+        """
+        Calculates the correlation function from the spectral density by
+        numeric integration, see Eq. 6 in the BoFiN paper
+        (DOI: 10.1103/PhysRevResearch.5.013181). This calculation requires that
+        the temperature of the bath (`self.T`) is specified.
+
+        Parameters
+        ----------
+        t : np.array or float
+            the time(s) at which to evaluate the correlation function
+        kwargs : will be passed to scipy's `quad_vec` function
+
+        Returns
+        -------
+            The correlation function at time t as an array or float.
+        """
+
+        def integrand(w, t):
+            return self.spectral_density(w) / np.pi * (
+                (2 * self._bose_einstein(w) + 1) * np.cos(w * t)
+                - 1j * np.sin(w * t)
+            )
+
+        result = quad_vec(lambda w: integrand(w, t), 0, np.Inf, **kwargs)
```

Comment:
AttributeError: `np.Inf` was removed in the NumPy 2.0 release. Use `np.inf` instead.